patches/server/0008-Paper-Metrics.patch

```diff
@@ -54,643 +67,140 @@ index 0000000000000000000000000000000000000000..e3b74dbdf8e14219a56fab939f3174e0
-+
-+        // Many servers tend to restart at a fixed time at xx:00 which causes an uneven distribution of requests on the
-+        // bStats backend. To circumvent this problem, we introduce some randomness into the initial and second delay.
-+        // WARNING: You must not modify any part of this Metrics class, including the submit delay or frequency!
-+        // WARNING: Modifying this code will get your plugin banned on bStats. Just don't do it!
-+        long initialDelay = (long) (1000 * 60 * (3 + Math.random() * 3));
-+        long secondDelay = (long) (1000 * 60 * (Math.random() * 30));
-+        scheduler.schedule(submitTask, initialDelay, TimeUnit.MILLISECONDS);
-+        scheduler.scheduleAtFixedRate(submitTask, initialDelay + secondDelay, 1000 * 60 * 30, TimeUnit.MILLISECONDS);
-+    }
-+
-+    /**
-+     * Gets the plugin specific data.
-+     *
-+     * @return The plugin specific data.
-+     */
-+    private JSONObject getPluginData() {
-+        JSONObject data = new JSONObject();
-+
-+        data.put("pluginName", name); // Append the name of the server software
-+        JSONArray customCharts = new JSONArray();
-+        for (CustomChart customChart : charts) {
-+            // Add the data of the custom charts
-+            JSONObject chart = customChart.getRequestJsonObject();
-+            if (chart == null) { // If the chart is null, we skip it
-+                continue;
-+            }
-+            customCharts.add(chart);
-+        }
-+        data.put("customCharts", customCharts);
-+
-+        return data;
-+    }
-+
-+    /**
-+     * Gets the server specific data.
-+     *
-+     * @return The server specific data.
-+     */
-+    private JSONObject getServerData() {
-+        // OS specific data
-+        String osName = System.getProperty("os.name");
-+        String osArch = System.getProperty("os.arch");
-+        String osVersion = System.getProperty("os.version");
-+        int coreCount = Runtime.getRuntime().availableProcessors();
-+
-+        JSONObject data = new JSONObject();
-+
-+        data.put("serverUUID", serverUUID);
++    private static Logger logger = LoggerFactory.getLogger("bStats");
 +
-+        data.put("osName", osName);
-+        data.put("osArch", osArch);
-+        data.put("osVersion", osVersion);
-+        data.put("coreCount", coreCount);
-+
-+        return data;
-+    }
-+
-+    /**
-+     * Collects the data and sends it afterwards.
-+     */
-+    private void submitData() {
-+        final JSONObject data = getServerData();
-+
-+        JSONArray pluginData = new JSONArray();
-+        pluginData.add(getPluginData());
-+        data.put("plugins", pluginData);
++    private MetricsBase metricsBase;
 +
++    public Metrics() {
++        File bStatsFolder = new File((File) MinecraftServer.getServer().options.valueOf("plugins"), "bStats");
++        File configFile = new File(bStatsFolder, "config.yml");
++        MetricsConfig config;
```

Comment:
Unfortunately, you cannot use the MetricsConfig. You still have to create and parse the config file yourself.
The `MetricsConfig` is not a YAML file but uses its own format that looks like this and is not compatible with the existing config files:
```
# bStats (https://bStats.org) collects some basic information for plugin authors, like
# how many people use their plugin and their total player count. It's recommended to keep
# bStats enabled, but if you're not comfortable with this, you can turn this setting off.
# There is no performance penalty associated with having metrics enabled, and data sent to
# bStats is fully anonymous.");
enabled=true
server-uuid=<UUID>
log-errors=false
log-sent-data=false
log-response-status-text=false
```
(see source: https://github.com/Bastian/bStats-Metrics/blob/master/base/src/main/java/org/bstats/config/MetricsConfig.java)

This config is only intended for new platforms and cannot be used by Paper for legacy reasons.
For the config, you should use the Bukkit Metrics class as a reference: https://github.com/Bastian/bStats-Metrics/blob/master/bukkit/src/main/java/org/bstats/bukkit/Metrics.java#L39-L57

Reply:
ahh, seeing that bukkit didn't use it confused me, that makes sense.
alternatively we could migrate existing files? ah no, cause other plugins, meh. ok, will fix later